<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="es" xml:lang="es"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="OMICs Analysis">

<title>Perfil Taxonómico de metagenomas usando Kraken</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/quarto.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="index_files/libs/quarto-html/anchor.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="index_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="index_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="index_files/libs/bootstrap/bootstrap-a312d171523933343ca77b5e293f9fd6.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<style>html{ scroll-behavior: smooth; }</style>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Contenido</h2>
   
  <ul>
  <li><a href="#introducción" id="toc-introducción" class="nav-link active" data-scroll-target="#introducción">Introducción</a>
  <ul class="collapse">
  <li><a href="#preparación-de-base-de-datos-indexada" id="toc-preparación-de-base-de-datos-indexada" class="nav-link" data-scroll-target="#preparación-de-base-de-datos-indexada">Preparación de base de datos indexada</a></li>
  </ul></li>
  <li><a href="#kraken2" id="toc-kraken2" class="nav-link" data-scroll-target="#kraken2">Kraken2</a>
  <ul class="collapse">
  <li><a href="#minusb" id="toc-minusb" class="nav-link" data-scroll-target="#minusb">MinusB</a></li>
  <li><a href="#standard-16gb" id="toc-standard-16gb" class="nav-link" data-scroll-target="#standard-16gb">Standard 16GB</a></li>
  <li><a href="#output-kraken2" id="toc-output-kraken2" class="nav-link" data-scroll-target="#output-kraken2">Output Kraken2</a>
  <ul class="collapse">
  <li><a href="#kraken2_output_.tsv" id="toc-kraken2_output_.tsv" class="nav-link" data-scroll-target="#kraken2_output_.tsv"><code>kraken2_output_*.tsv</code></a></li>
  <li><a href="#kraken2_report_.txt" id="toc-kraken2_report_.txt" class="nav-link" data-scroll-target="#kraken2_report_.txt"><code>kraken2_report_*.txt</code></a></li>
  <li><a href="#kraken2_classified_.fq" id="toc-kraken2_classified_.fq" class="nav-link" data-scroll-target="#kraken2_classified_.fq"><code>kraken2_classified_*.fq</code></a></li>
  <li><a href="#kraken2_unclassified_.fq" id="toc-kraken2_unclassified_.fq" class="nav-link" data-scroll-target="#kraken2_unclassified_.fq"><code>kraken2_unclassified_*.fq</code></a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Perfil Taxonómico de metagenomas usando Kraken</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Autor/a</div>
    <div class="quarto-title-meta-contents">
             <p>OMICs Analysis </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>

<nav id="TOC-body" role="doc-toc">
    <h2 id="toc-title">Contenido</h2>
   
  <ul>
  <li><a href="#introducción" id="toc-introducción">Introducción</a>
  <ul>
  <li><a href="#preparación-de-base-de-datos-indexada" id="toc-preparación-de-base-de-datos-indexada">Preparación de base de datos indexada</a></li>
  </ul></li>
  <li><a href="#kraken2" id="toc-kraken2">Kraken2</a>
  <ul>
  <li><a href="#minusb" id="toc-minusb">MinusB</a></li>
  <li><a href="#standard-16gb" id="toc-standard-16gb">Standard 16GB</a></li>
  <li><a href="#output-kraken2" id="toc-output-kraken2">Output Kraken2</a>
  <ul>
  <li><a href="#kraken2_output_.tsv" id="toc-kraken2_output_.tsv"><code>kraken2_output_*.tsv</code></a></li>
  <li><a href="#kraken2_report_.txt" id="toc-kraken2_report_.txt"><code>kraken2_report_*.txt</code></a></li>
  <li><a href="#kraken2_classified_.fq" id="toc-kraken2_classified_.fq"><code>kraken2_classified_*.fq</code></a></li>
  <li><a href="#kraken2_unclassified_.fq" id="toc-kraken2_unclassified_.fq"><code>kraken2_unclassified_*.fq</code></a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/logo_OAN_for_white.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="300"></p>
</figure>
</div>
<section id="introducción" class="level1">
<h1>Introducción</h1>
<p>Vamos a utilizar <code>Kraken v2</code>, lo puedes descragar <a href="https://ccb.jhu.edu/software/kraken/MANUAL.html">aquí</a>.</p>
<p>Este tutorial se hizo utilizando un equipo A<code>SUSTeK COMPUTER INC. G75VW</code> con procesador <code>Intel® Core™ i7-3610QM CPU @ 2.30GHz × 8</code>, <code>24 GB</code> de <code>RAM</code>.</p>
<p>Advertencia ⚠️. Según tus capacidades computacionales puedes utilizar bases de datos más grandes para clasificar, <strong>entre mayor tamaño mayor cantidad de memoría <code>RAM</code> necesitarás</strong>. Aquí te dejo algunas disponibles, puedes ver más a detalle de su indexación y la gran variedad de bases de datos que ya existen <a href="https://benlangmead.github.io/aws-indexes/k2">aquí</a>:</p>
<table class="caption-top table" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Colección</th>
<th data-quarto-table-cell-role="th">Contenido</th>
<th data-quarto-table-cell-role="th">Fecha</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Tamaño de archivo (GB)</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Tamaño Index (GB)</th>
<th data-quarto-table-cell-role="th">HTTPS URL</th>
<th data-quarto-table-cell-role="th">Inspeccionar</th>
<th data-quarto-table-cell-role="th">Librería</th>
<th data-quarto-table-cell-role="th">MD5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Viral</td>
<td>Refseq viral</td>
<td>7/14/2025</td>
<td style="text-align: right;">0.5</td>
<td style="text-align: right;">0.6</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_viral_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/viral_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/viral_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/viral_20250714/viral.md5">.md5</a></td>
</tr>
<tr class="even">
<td>MinusB</td>
<td>Refseq archaea, viral, plasmid, human<sup>1</sup>, UniVec_Core</td>
<td>7/14/2025</td>
<td style="text-align: right;">7.7</td>
<td style="text-align: right;">10.8</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_minusb_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/minusb_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/minusb_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/minusb_20250714/minusb.md5">.md5</a></td>
</tr>
<tr class="odd">
<td>Standard</td>
<td>Refseq archaea, bacteria, viral, plasmid, human<sup>1</sup>, UniVec_Core</td>
<td>7/14/2025</td>
<td style="text-align: right;">70.8</td>
<td style="text-align: right;">91.8</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_standard_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_20250714/standard.md5">.md5</a></td>
</tr>
<tr class="even">
<td>Standard-8</td>
<td>Standard with DB capped at 8 GB</td>
<td>7/14/2025</td>
<td style="text-align: right;">5.5</td>
<td style="text-align: right;">7.5</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_standard_08_GB_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_08_GB_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_08_GB_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_08_GB_20250714/standard_08_GB.md5">.md5</a></td>
</tr>
<tr class="odd">
<td>Standard-16</td>
<td>Standard with DB capped at 16 GB</td>
<td>7/14/2025</td>
<td style="text-align: right;">11.2</td>
<td style="text-align: right;">14.9</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_standard_16_GB_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_16_GB_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_16_GB_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/standard_16_GB_20250714/standard_16_GB.md5">.md5</a></td>
</tr>
<tr class="even">
<td>PlusPF</td>
<td>Standard plus Refseq protozoa &amp; fungi</td>
<td>7/14/2025</td>
<td style="text-align: right;">75.9</td>
<td style="text-align: right;">98.4</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_pluspf_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_20250714/pluspf.md5">.md5</a></td>
</tr>
<tr class="odd">
<td>PlusPF-8</td>
<td>PlusPF with DB capped at 8 GB</td>
<td>7/14/2025</td>
<td style="text-align: right;">5.5</td>
<td style="text-align: right;">7.5</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_pluspf_08_GB_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_08_GB_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_08_GB_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_08_GB_20250714/pluspf_08_GB.md5">.md5</a></td>
</tr>
<tr class="even">
<td>PlusPF-16</td>
<td>PlusPF with DB capped at 16 GB</td>
<td>7/14/2025</td>
<td style="text-align: right;">11.2</td>
<td style="text-align: right;">14.9</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_pluspf_16_GB_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_16_GB_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_16_GB_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspf_16_GB_20250714/pluspf_16_GB.md5">.md5</a></td>
</tr>
<tr class="odd">
<td>PlusPFP</td>
<td>Standard plus Refseq protozoa, fungi &amp; plant</td>
<td>7/14/2025</td>
<td style="text-align: right;">155.7</td>
<td style="text-align: right;">210.6</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_pluspfp_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_20250714/pluspfp.md5">.md5</a></td>
</tr>
<tr class="even">
<td>PlusPFP-8</td>
<td>PlusPFP with DB capped at 8 GB</td>
<td>7/14/2025</td>
<td style="text-align: right;">5.2</td>
<td style="text-align: right;">7.5</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_pluspfp_08_GB_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_08_GB_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_08_GB_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_08_GB_20250714/pluspfp_08_GB.md5">.md5</a></td>
</tr>
<tr class="odd">
<td>PlusPFP-16</td>
<td>PlusPFP with DB capped at 16 GB</td>
<td>7/14/2025</td>
<td style="text-align: right;">10.5</td>
<td style="text-align: right;">14.9</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_pluspfp_16_GB_20250714.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_16_GB_20250714/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_16_GB_20250714/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/pluspfp_16_GB_20250714/pluspfp_16_GB.md5">.md5</a></td>
</tr>
<tr class="even">
<td>EuPathDB46<sup>2</sup></td>
<td>Eukaryotic pathogen genomes with contaminants removed</td>
<td>4/18/2023</td>
<td style="text-align: right;">8.4</td>
<td style="text-align: right;">11</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_eupathdb48_20230407.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_eupathdb48_20230407/kraken2inspect_output.txt">.txt</a></td>
<td>N/A</td>
<td>N/A</td>
</tr>
<tr class="odd">
<td>core_nt Database</td>
<td>Very large collection, inclusive of GenBank, RefSeq, TPA and PDB</td>
<td>6/9/2025</td>
<td style="text-align: right;">183.0</td>
<td style="text-align: right;">238.2</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_core_nt_20250609.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/core_nt_20250609/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/core_nt_20250609/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/core_nt_20250609/core_nt.md5">.md5</a></td>
</tr>
<tr class="even">
<td>GTDB v226 (<code class="language-plaintext highlighter-rouge">genomic_files_reps</code>)</td>
<td>Bacterial and archaeal</td>
<td>6/9/2025</td>
<td style="text-align: right;">500.9</td>
<td style="text-align: right;">644.0</td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/k2_gtdb_genome_reps_20250609.tar.gz">.tar.gz</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/gtdb_genome_reps_20250609/inspect.txt">.txt</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/gtdb_genome_reps_20250609/library_report.tsv">.tsv</a></td>
<td><a href="https://genome-idx.s3.amazonaws.com/kraken/gtdb_genome_reps_20250609/gtdb_genome_reps.md5">.md5</a></td>
</tr>
</tbody>
</table>
<p>Para este tutorial vamos a trabajar con la base de datos indexada <code>MinusB</code> (enfocada en bacterias y arqueas, más ligera) y <code>Standard-16</code> (más completa, incluyendo eucariotas, pero requiere más memoria). Como datos de metagenoma vamos a utilizar los que previamente habiamos trimeado y filtrado de contaminación de hospedero.</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>SRA ID</th>
<th>Condición</th>
<th>Etapa de crecimiento</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR19746212">SRR19746212</a></td>
<td>Periodo Normal</td>
<td>Novillas en crecimiento</td>
</tr>
<tr class="even">
<td><a href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR19746215">SRR19746215</a></td>
<td>Estrés Térmico</td>
<td>Novillas en crecimiento</td>
</tr>
<tr class="odd">
<td><a href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR19746202">SRR19746202</a></td>
<td>Periodo Normal</td>
<td>Novillas</td>
</tr>
<tr class="even">
<td><a href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR19746205">SRR19746205</a></td>
<td>Estrés Térmico</td>
<td>Novillas</td>
</tr>
<tr class="odd">
<td><a href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR19746208">SRR19746208</a></td>
<td>Periodo Normal</td>
<td>Vacas Lactantes</td>
</tr>
<tr class="even">
<td><a href="https://trace.ncbi.nlm.nih.gov/Traces/sra?run=SRR19746219">SRR19746219</a></td>
<td>Estrés Térmico</td>
<td>Vacas Lactantes</td>
</tr>
</tbody>
</table>
<section id="preparación-de-base-de-datos-indexada" class="level2">
<h2 class="anchored" data-anchor-id="preparación-de-base-de-datos-indexada">Preparación de base de datos indexada</h2>
<p>Para este proceso vamos a generar un directorio llamado <code>2025_metagenomics_profile</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 2025_metagenomics_profile</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 2025_metagenomics_profile</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> dbs</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> dbs</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> db_minusb</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> db_minusb</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://genome-idx.s3.amazonaws.com/kraken/k2_minusb_20250714.tar.gz</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xzf</span> k2_minusb_20250714.tar.gz</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> db_standard</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> db_standard</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://genome-idx.s3.amazonaws.com/kraken/k2_standard_16_GB_20250714.tar.gz</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xzf</span> k2_standard_16_GB_20250714.tar.gz</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="kraken2" class="level1">
<h1>Kraken2</h1>
<p><code>Kraken2</code> es una herramienta bioinformática rápida y precisa para clasificar secuencias de ADN de muestras metagenómicas (como las de microbiomas) en categorías taxonómicas, como bacterias, virus, arqueas o eucariotas. Es ampliamente usada en estudios de microbiología y ecología para identificar “quién está ahí” en una mezcla de organismos.</p>
<p><strong>Cómo funciona de manera simple</strong>:</p>
<ul>
<li><p><strong>Construcción de la base de datos</strong>: Kraken2 crea un índice (database) a partir de genomas de referencia (como los que descargaste: minusb para bacterias y arqueas, o standard para una cobertura más amplia). Divide los genomas en “k-mers” (subsecuencias cortas de longitud fija, por ejemplo, 31 bases). Cada k-mer se asocia al ancestro común más bajo (LCA, por sus siglas en inglés) de los taxones que lo comparten, para manejar ambigüedades.</p></li>
<li><p><strong>Clasificación de reads</strong>: Para tus archivos FASTQ (secuencias de lecturas), Kraken2 divide cada read en k-mers y los busca en la base de datos. Cuenta los “hits” (coincidencias) para diferentes taxones y asigna el read al taxón con el mejor puntaje (basado en la raíz del árbol taxonómico). Si no hay coincidencias, el read se marca como “no clasificado”.</p></li>
<li><p><strong>Ventajas</strong>: Es muy rápido (usa hashing para búsquedas eficientes) y maneja datos paired-end como los tuyos. Puedes ajustar parámetros como la confianza (–confidence) para mayor precisión.</p></li>
<li><p><strong>Limitaciones</strong>: Depende de la calidad de la base de datos; si un organismo no está en ella, no se detectará.</p></li>
</ul>
<p>Para este punto es conveniente tener las secuencias pareadas en un solo archivo concatenado FASTQ. En caso de que tengas tus archivos en dos archivos separados <code>fwd</code> y <code>rev</code>, puedes preprocesarlos usando <code>kraken --paired</code>.</p>
<p>Explicación breve de opciones usadas:</p>
<ul>
<li><p><code>--paired</code> : entrada paired-end.</p></li>
<li><p><code>--gzip-compressed</code> : tus fastq están gz.</p></li>
<li><p><code>--use-names</code> : muestra nombres taxonómicos legibles en el output.</p></li>
<li><p><code>--report</code> : resumen por taxón (ideal para downstream).</p></li>
<li><p><code>--output</code> : asignaciones por lectura (format columnas).</p></li>
<li><p><code>--classified-out</code> / <code>--unclassified-out</code> : opcional, guarda reads clasificados/no clasificados (usar <code>#</code> para pares).</p></li>
<li><p><code>--confidence</code> : opcional, umbral de confianza (ejemplo <code>0.1</code> si quieres mayor especificidad).</p></li>
</ul>
<section id="minusb" class="level2">
<h2 class="anchored" data-anchor-id="minusb">MinusB</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="st">"./bin/kraken2/kraken2"</span> <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--db</span> <span class="st">"dbs/db_minusb"</span><span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--paired</span> <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--gzip-compressed</span> <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--threads</span> 7 <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--use-names</span> <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report</span> <span class="st">"kraken2_report_k2_minusb_20250714.txt"</span> <span class="dt">\</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--output</span> <span class="st">"kraken2_output_k2_minusb_20250714.tsv"</span> <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--classified-out</span> <span class="st">"kraken2_classified_k2_minusb_20250714_#.fq"</span> <span class="dt">\</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--unclassified-out</span> <span class="st">"kraken2_unclassified_k2_minusb_20250714_#.fq"</span> <span class="dt">\</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"1_QC/host_filter/SRR19746202_1.fastq.gz"</span> <span class="st">"1_QC/host_filter/SRR19746202_2.fastq.gz"</span></span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Resultado:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">32842930</span> sequences <span class="er">(</span><span class="ex">9852.88</span> Mbp<span class="kw">)</span> <span class="ex">processed</span> in 751.006s <span class="er">(</span><span class="ex">2623.9</span> Kseq/m, 787.17 Mbp/m<span class="kw">)</span><span class="bu">.</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="ex">3196540</span> sequences classified <span class="er">(</span><span class="ex">9.73%</span><span class="kw">)</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">29646390</span> sequences unclassified <span class="er">(</span><span class="ex">90.27%</span><span class="kw">)</span></span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="standard-16gb" class="level2">
<h2 class="anchored" data-anchor-id="standard-16gb">Standard 16GB</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="st">"./bin/kraken2/kraken2"</span> <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--db</span> <span class="st">"dbs/db_standard"</span><span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--paired</span> <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--gzip-compressed</span> <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--threads</span> 7 <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--use-names</span> <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report</span> <span class="st">"kraken2_report_k2_standard_20250714.txt"</span> <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--output</span> <span class="st">"kraken2_output_k2_standard_20250714.tsv"</span> <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--classified-out</span> <span class="st">"kraken2_classified_k2_standard_20250714_#.fq"</span> <span class="dt">\</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--unclassified-out</span> <span class="st">"kraken2_unclassified_k2_standard_20250714_#.fq"</span> <span class="dt">\</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"1_QC/host_filter/SRR19746202_1.fastq.gz"</span> <span class="st">"1_QC/host_filter/SRR19746202_2.fastq.gz"</span></span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Resultado:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">32842930</span> sequences <span class="er">(</span><span class="ex">9852.88</span> Mbp<span class="kw">)</span> <span class="ex">processed</span> in 581.113s <span class="er">(</span><span class="ex">3391.0</span> Kseq/m, 1017.31 Mbp/m<span class="kw">)</span><span class="bu">.</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="ex">9536390</span> sequences classified <span class="er">(</span><span class="ex">29.04%</span><span class="kw">)</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">23306540</span> sequences unclassified <span class="er">(</span><span class="ex">70.96%</span><span class="kw">)</span></span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Escoger la base de datos adecuada influye drasticamente en la clasificación de las lecturas.</p>
</section>
<section id="output-kraken2" class="level2">
<h2 class="anchored" data-anchor-id="output-kraken2">Output Kraken2</h2>
<p><code>Kraken2</code> nos arroja como <code>output</code> los siguientes archivos:</p>
<p>Cuando corres los comandos que te pasé, obtienes <strong>cuatro tipos principales de archivos por base de datos</strong>:</p>
<section id="kraken2_output_.tsv" class="level3">
<h3 class="anchored" data-anchor-id="kraken2_output_.tsv"><code>kraken2_output_*.tsv</code></h3>
<ul>
<li><p>Contiene una línea <strong>por cada lectura</strong>.</p>
<p>Columnas típicas:</p>
<ol type="1">
<li><p>C/U → si fue clasificada (C) o no (U).</p></li>
<li><p>Read ID.</p></li>
<li><p>TaxID asignado.</p></li>
<li><p>Longitud de la secuencia.</p></li>
<li><p>K-mers clasificados / asignaciones.</p></li>
</ol></li>
<li><p>Útil si quieres rastrear lecturas individuales, pero <strong>no es práctico para gráficos de composición</strong>.</p></li>
</ul>
</section>
<section id="kraken2_report_.txt" class="level3">
<h3 class="anchored" data-anchor-id="kraken2_report_.txt"><code>kraken2_report_*.txt</code></h3>
<ul>
<li><p>Es el archivo <strong>más importante para resúmenes</strong>.</p>
<p>Contiene la composición taxonómica del metagenoma en forma de tabla jerárquica.</p>
<p>Columnas típicas:</p>
<ol type="1">
<li><p><strong>% de lecturas</strong> asignadas a ese taxón (incluyendo descendientes).</p></li>
<li><p>lecturas asignadas directamente a ese taxón.</p></li>
<li><p>lecturas totales (taxón + descendientes).</p></li>
<li><p>TaxID.</p></li>
<li><p>Nivel taxonómico (U = root, D = dominio, P = filo, C = clase, O = orden, F = familia, G = género, S = especie, etc.).</p></li>
<li><p>Nombre del taxón.</p></li>
</ol></li>
<li><p>Este archivo <strong>sirve como entrada ideal</strong> para hacer gráficos de barras apiladas (ejemplo: abundancias relativas por filo o por género).</p></li>
<li><p>Si después usas <strong>Bracken</strong>, este mismo archivo puede refinarse para obtener abundancias corregidas.</p></li>
</ul>
</section>
<section id="kraken2_classified_.fq" class="level3">
<h3 class="anchored" data-anchor-id="kraken2_classified_.fq"><code>kraken2_classified_*.fq</code></h3>
<ul>
<li>Contiene las lecturas que <strong>sí fueron clasificadas</strong> por Kraken2, separadas en pares (<code>_1.fq</code> y <code>_2.fq</code>).</li>
</ul>
<p>Útil si quieres analizar solo la fracción clasificada (ej. ensamblaje por taxón específico).</p>
</section>
<section id="kraken2_unclassified_.fq" class="level3">
<h3 class="anchored" data-anchor-id="kraken2_unclassified_.fq"><code>kraken2_unclassified_*.fq</code></h3>
<ul>
<li><p>Contiene las lecturas que <strong>no pudieron clasificarse</strong> (no coincidieron con la base de datos).</p></li>
<li><p>Útil si quieres estudiar la “parte desconocida” del metagenoma o reanalizar con otra base de datos.</p></li>
</ul>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copiado");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copiado");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<script src="index_files/libs/quarto-html/zenscroll-min.js"></script>
</body></html>